### ISSUE TITLE
Feature: Voice Identification (Speaker Recognition)

### ISSUE THREAD

**Issue by whatiskadudoing** [whatiskadudoing/koe#23]
## Overview

Identify speakers by their voice using speaker embeddings. Instead of "Speaker 1, Speaker 2", recognize specific people: "John", "Sarah", etc.

## Why This Matters

- **Personalized transcripts** — See actual names, not generic labels
- **Recurring meeting attendees** — Automatically recognize team members
- **Voice profiles** — Build a personal voice directory over time

## How It Works

```
┌─────────────────────────────────────────────────────────────────┐
│                    VOICE IDENTIFICATION FLOW                    │
├─────────────────────────────────────────────────────────────────┤
│                                                                 │
│  ENROLLMENT (One-time per person)                               │
│  ┌─────────────────────────────────────┐                        │
│  │  Record 5-10 seconds of speech      │                        │
│  │           ↓                         │                        │
│  │  ECAPA-TDNN Neural Network          │                        │
│  │           ↓                         │                        │
│  │  192-dimensional embedding          │                        │
│  │  [0.12, -0.34, 0.56, ..., 0.78]    │                        │
│  │           ↓                         │                        │
│  │  Save as "John's voice profile"     │                        │
│  └─────────────────────────────────────┘                        │
│                                                                 │
│  RECOGNITION (During transcription)                             │
│  ┌─────────────────────────────────────┐                        │
│  │  New speaker segment                │                        │
│  │           ↓                         │                        │
│  │  Extract embedding                  │                        │
│  │           ↓                         │                        │
│  │  Compare with saved profiles        │                        │
│  │  (Cosine Similarity)                │                        │
│  │           ↓                         │                        │
│  │  Score > 0.7? → "John"              │                        │
│  │  Score < 0.7? → "Unknown Speaker"   │                        │
│  └─────────────────────────────────────┘                        │
│                                                                 │
└─────────────────────────────────────────────────────────────────┘
```

### Speaker Embeddings
A speaker embedding is a fixed-size vector that represents the unique characteristics of someone's voice:
- Voice pitch patterns
- Speaking rhythm
- Vocal tract characteristics
- Pronunciation patterns

### ECAPA-TDNN Architecture
State-of-the-art model for speaker verification:
- Time Delay Neural Network with attention
- Trained on VoxCeleb dataset (7000+ speakers)
- Output: 192-dimensional embedding vector
- Comparison: Cosine similarity between embeddings

### Option 1: FluidAudio (Recommended)
Native Swift SDK with speaker embeddings on Apple Neural Engine.

```swift
import FluidAudio

// Extract speaker embedding
let embedding = try await speakerEmbedding.embed(audioURL: sampleURL)

// Compare two voice samples
let similarity = cosineSimilarity(embedding1, embedding2)
let isSameSpeaker = similarity > 0.7
```

**Resources:**
- [FluidAudio GitHub](https://github.com/FluidInference/FluidAudio)
- [FluidAudio Blog](https://iosdev.tools/blog/fluidaudio/)

### Option 2: SpeechBrain + Core ML Conversion
Convert Python models to Core ML for on-device use.

```python
# 1. Export SpeechBrain model to ONNX
from speechbrain.inference.speaker import SpeakerRecognition
model = SpeakerRecognition.from_hparams(source="speechbrain/spkrec-ecapa-voxceleb")

# 2. Convert to Core ML
import coremltools as ct
mlmodel = ct.convert(onnx_model)
mlmodel.save("SpeakerEmbedding.mlmodel")
```

```swift
// 3. Use in Swift
let model = try SpeakerEmbedding(configuration: .init())
let embedding = try model.prediction(audio: audioBuffer)
```

**Resources:**
- [SpeechBrain ECAPA-TDNN](https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb)
- [Core ML Tools](https://coremltools.readme.io/)

### Option 3: Sherpa-ONNX
Cross-platform ONNX runtime approach.

```swift
// Using Sherpa-ONNX speaker embedding
let config = SpeakerEmbeddingConfig(modelPath: "wespeaker.onnx")
let extractor = SpeakerEmbeddingExtractor(config: config)
let embedding = extractor.compute(samples: audioSamples)
```

**Resources:**
- [Sherpa-ONNX](https://github.com/k2-fsa/sherpa-onnx)
- [WeSpeaker Models](https://github.com/wenet-e2e/wespeaker)

### Data Structure
```swift
struct VoiceProfile: Codable {
    let id: UUID
    let name: String
    let embedding: [Float]  // 192-dimensional vector
    let createdAt: Date
    let sampleCount: Int
}

class VoiceProfileManager {
    private var profiles: [VoiceProfile] = []
    
    // Enroll a new voice
    func enroll(name: String, audioSamples: [URL]) async throws -> VoiceProfile {
        // Extract embeddings from multiple samples
        var embeddings: [[Float]] = []
        for sample in audioSamples {
            let embedding = try await extractEmbedding(from: sample)
            embeddings.append(embedding)
        }
        
        // Average embeddings for robustness
        let averageEmbedding = averageVectors(embeddings)
        
        return VoiceProfile(
            id: UUID(),
            name: name,
            embedding: averageEmbedding,
            createdAt: Date(),
            sampleCount: audioSamples.count
        )
    }
    
    // Identify speaker from audio
    func identify(audioURL: URL, threshold: Float = 0.7) async throws -> String? {
        let embedding = try await extractEmbedding(from: audioURL)
        
        var bestMatch: (name: String, score: Float)?
        for profile in profiles {
            let score = cosineSimilarity(embedding, profile.embedding)
            if score > threshold {
                if bestMatch == nil || score > bestMatch!.score {
                    bestMatch = (profile.name, score)
                }
            }
        }
        
        return bestMatch?.name
    }
}
```

### Cosine Similarity
```swift
func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
    guard a.count == b.count else { return 0 }
    
    var dotProduct: Float = 0
    var normA: Float = 0
    var normB: Float = 0
    
    for i in 0..<a.count {
        dotProduct += a[i] * b[i]
        normA += a[i] * a[i]
        normB += b[i] * b[i]
    }
    
    return dotProduct / (sqrt(normA) * sqrt(normB))
}
```

### Enrollment Flow
1. User taps "Add Voice Profile"
2. Prompted to say a few sentences (10-15 seconds)
3. App extracts embedding and saves with name
4. Optional: Add more samples for better accuracy

### Recognition Integration
```
Before: [Speaker 1] Let's discuss the quarterly results.
        [Speaker 2] I'll share my screen.

After:  [John] Let's discuss the quarterly results.
        [Sarah] I'll share my screen.
```

### Settings
- [ ] Manage voice profiles (add/edit/delete)
- [ ] Confidence threshold slider
- [ ] Auto-suggest profile from contacts
- [ ] Export/import voice profiles

### Factors Affecting Recognition
- **Audio Quality** — Background noise reduces accuracy
- **Sample Duration** — More samples = better profile
- **Recording Conditions** — Phone vs. laptop mic
- **Speaker Variability** — Sick voice, different mood

### Best Practices
1. Enroll with 3-5 samples in different conditions
2. Use at least 5 seconds of speech per sample
3. Re-enroll if recognition degrades
4. Combine with diarization for multi-speaker scenarios

### Models
- [SpeechBrain ECAPA-TDNN](https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb)
- [WeSpeaker](https://github.com/wenet-e2e/wespeaker)
- [FluidAudio Models](https://github.com/FluidInference/FluidAudio)

### Research
- [ECAPA-TDNN Paper](https://arxiv.org/abs/2005.07143)
- [VoxCeleb Dataset](https://www.robots.ox.ac.uk/~vgg/data/voxceleb/)

### Tutorials
- [Speaker Verification with SpeechBrain](https://speechbrain.github.io/)
- [Voice-Based Security Implementation](https://medium.com/@rudderanalytics/voice-based-security-implementing-a-robust-speaker-verification-system-12c5fd98f1c1)

## Implementation Steps

1. **Research Phase**
   - [ ] Test FluidAudio speaker embedding API
   - [ ] Evaluate accuracy with test recordings
   - [ ] Benchmark embedding extraction speed

2. **Core Implementation**
   - [ ] Integrate speaker embedding model
   - [ ] Build voice profile storage (Core Data/SQLite)
   - [ ] Implement cosine similarity matching

3. **User Experience**
   - [ ] Design enrollment UI
   - [ ] Add profile management screen
   - [ ] Integrate with transcription view

4. **Optimization**
   - [ ] Cache embeddings for faster lookup
   - [ ] Batch processing for efficiency
   - [ ] Profile update/refinement over time

## Privacy Considerations

- Voice profiles stored locally only
- Never upload embeddings to cloud
- User can delete profiles at any time
- Clear consent for voice enrollment

## Labels

`enhancement` `ml` `speaker-recognition` `privacy` `help wanted`

**Comment by whatiskadudoing** [comment]
## Additional Research Findings

### Voice Identification Implementation Patterns

Research into professional audio processing applications revealed important voice identification patterns:

#### 1. Training Data Collection

Professional apps collect training data for voice profiles:

```javascript
AudioService.uploadTrainingData({
  userId: 12935226,
  audioFile: "training_sample.wav",
  duration: 30  // seconds
})

// Handle short samples
"failed to remove short training data file"
// Minimum duration required for good embeddings
```

#### 2. Accent Voice Selection

Voice profiles can include accent preferences:

```json
{
  "SELECTED_ACCENT_VOICE_NAME": "alan",
  "USER_ACCENT_SET": true,
  "source_accent": "indian"
}
```

Available accent categories observed:
- Indian
- Filipino  
- Latin American (LatAm)
- African
- Inbound (for processing incoming audio)

#### 3. Speaker Capture Modes

Different capture modes for voice identification:

```javascript
captureType: {
  CAPTURE_MIC_RTC_ASR: 1,      // Real-time ASR from mic
  CAPTURE_SPK_TRAINING_REC: 2  // Training recording from speaker
}
```

#### 4. Energy Level Tracking

Track voice energy for speaker activity:

```javascript
{
  currentSpkEnergyLevel: 0.75,
  frameEnergy: 0.82,
  mic_voice_level: "medium"
}
```

Energy levels help with:
- Voice Activity Detection (VAD)
- Speaker change detection
- Quality assessment

#### 5. Word-Level Processing

For accurate speaker identification, track word boundaries:

```javascript
{
  wordsInbound: [...],      // Words from remote speakers
  wordsWindowSize: 50,      // Sliding window for analysis
  piiIndexesInbound: [...]  // PII locations in transcript
}
```

#### 6. Transcript Encryption

Voice identification data and transcripts are encrypted:

```javascript
// calltranscript.json contains encrypted hex strings
// Data encrypted at rest for privacy
{
  callId: "cd3c4dc5-186a-4ec1-b0ca-abaa5f3e2656",
  data: "3a14dd807edef1043745881038e1bfa79a4be612..."  // AES encrypted
}
```

#### 7. Session Management

Track voice identification sessions:

```javascript
{
  vtSessionTimeout: 30000,     // Voice translation session timeout
  sessionId: "...",
  streamId: "..."
}
```

#### 8. Model Selection Based on Device

Different voice models for different audio quality:

```javascript
// High-quality devices get better models
if (deviceAllowList.includes(currentDevice)) {
  useHighQualityModel()
} else {
  useStandardModel()
}
```

Device allowlist includes premium audio devices (AirPods, Jabra, etc.)
